File: PoC - PuzzleGeneration/sudoku_generator.py
import random
import copy

class SudokuGenerator: 
  # initialising the grid to an empty 9x9 grid
  def __init__(self, level='Easy'):
    self.grid = [[0 for i in range(9)] for j in range(9)]
    self.level = level
    self.difficulty_levels = {
        'Easy': random.randint(31, 35),
        'Medium': random.randint(35, 39),
        'Hard': random.randint(39, 43),
        'Expert': random.randint(46, 50)
    }
  
  # reusable method to reinitialise an empty state of the grid
  def initialise_grid(self):
    self.grid = [[0 for i in range(9)] for j in range(9)]

  # method to fill the grid with a complete solution
  def is_valid_placement(self, grid, num, pos):
    # check row for the presence of the same number
    for i in range(9):
      if grid[pos[0]][i] == num and pos[1] != i:
        return False

    # check column for the presence of the same number
    for i in range(9):
      if grid[i][pos[1]] == num and pos[0] != i:
        return False
    
    # calculate subgrid start points
    box_start_row = pos[0] - pos[0] % 3
    box_start_col = pos[1] - pos[1] % 3

    # check 3x3 subgrid for the presence of the same number
    for i in range(box_start_row, box_start_row + 3):
      for j in range(box_start_col, box_start_col + 3):
        if grid[i][j] == num and (i, j) != pos:
          return False
    
    # num is not present in the row, column or subgrid = valid placement
    return True 
  
  # recursive function to randomly fill the grid
  def recursive_fill_grid(self, grid=None):
    if grid is None:
      grid = self.grid
    for i in range(9):
      for j in range(9):
        if grid[i][j] == 0:
          
          numbers = list(range(1,10)) # shuffles the numbers before trying them
          random.shuffle(numbers)
          for num in numbers:
            if self.is_valid_placement(grid, num, (i,j)): # attempts to place the number
              grid[i][j] = num
              if self.recursive_fill_grid(grid):
                return True
              grid[i][j] = 0
          return False
    return True
  
  # method to generate a complete grid
  def generate_complete_grid(self):
    self.initialise_grid()
    if not self.recursive_fill_grid():
      raise Exception('Failed to generate a complete Sudoku grid')
    
  # gets the coordinates of the non-empty squares
  def get_non_empty_squares(self):
    return [(row, col) for row in range(9) for col in range(9) if self.grid[row][col] != 0]

  # method to remove numbers from the grid to create the puzzle
  def remove_numbers_from_grid(self):
    grid = self.grid
    target_removed = self.difficulty_levels.get(self.level, 36) # default to easy level
    removed_squares = 0

    while removed_squares < target_removed:
      row, col = random.randint(0,8), random.randint(0,8)
      if grid[row][col] != 0:
        backup = grid[row][col]
        grid [row][col] = 0
        if not self.has_unique_solution(copy.deepcopy(grid)):
          grid[row][col] = backup # restore if the solution is not unique
        else:
          removed_squares += 1

    print(f"Removed {removed_squares} squares for a {self.level} puzzle.")


  # method to check if the puzzle has a unique solution
  def has_unique_solution(self, grid):
    solutions = [0]
    
    def solve(grid, non_empty_squares=None):
      if non_empty_squares is None:
        non_empty_squares = self.get_non_empty_squares()
            
      if not non_empty_squares:  # if there are no more non-empty squares, a solution has been found
        solutions[0] += 1
        return solutions[0] <= 1  # return False immediately if more than one solution is found
        
      for num in range(1, 10):
        row, col = non_empty_squares[0]  # work on the first non-empty square
        if self.is_valid_placement(grid, num, (row, col)):
          grid[row][col] = num  # try this number
          if solve(grid, non_empty_squares[1:]):  # move to the next square
            return True  # found a valid solution, stop the recursion
          grid[row][col] = 0  # reset the square if the number doesn't lead to a solution
  
      return False  # no valid number was found for the current square, backtrack
    
    solve(grid)
    return solutions[0] == 1
  
  # function to print the grid in the command line
  def print_grid(self, message="Grid"):
    print(f"{message}:")
    for row in range(9):
      if row % 3 == 0 and row != 0:
        print("- - - - - - - - - - - -")
      for col in range(9):
        if col % 3 == 0 and col != 0:
          print(" | ", end="")
        if col == 8:
          print(self.grid[row][col])
        else:
          print(str(self.grid[row][col]) + " ", end="")
    print("\n")

  # returns a copy of the grid to be used in solver.py
  def get_grid(self):
    return copy.deepcopy(self.grid)

def main():
  sudoku = SudokuGenerator(level="Hard")
  sudoku.generate_complete_grid()
  sudoku.print_grid(message = "Fully Generated Grid")
  
  sudoku.remove_numbers_from_grid()
  sudoku.print_grid(message = "Puzzle Grid")

  if sudoku.has_unique_solution(copy.deepcopy(sudoku.grid)):
    print("A unique solution exists.\n")
  else:
    print("No unique solution exists.")

if __name__ == "__main__": 
  main()

File: FinalCode/sudoku-app/algorithms/sudokuGenerator.js
// algorithm to generate a Sudoku puzzle:
// - based on a chosen difficulty level
// - checking if the generated puzzle has one unique solution
class SudokuGenerator {
  // set default level to "Easy"
  constructor(level = "Easy") {
    this.grid = Array.from({ length: 9 }, () => Array(9).fill(0)); // 9x9 grid with 0s
    this.level = level; // set the level
    this.difficultyLevels = {
      // different difficulty levels with number of removed clues for each level
      Easy: 40 + Math.floor(Math.random() * 4), // 40 to 43, Easy level
      Medium: 44 + Math.floor(Math.random() * 4), // 44 to 47, Medium level
      Hard: 48 + Math.floor(Math.random() * 4), // 48 to 51, Hard level
      Expert: 52 + Math.floor(Math.random() * 4), // 52 to 55, Expert level
    };
  }

  initialiseGrid() {
    // create a 9x9 grid with 0s
    this.grid = Array.from({ length: 9 }, () => Array(9).fill(0));
  }

  // function to check if the placement of a number is valid, Sudoku rules
  isValidPlacement(grid, num, pos) {
    // check if the number is already in the row or column
    for (let i = 0; i < 9; i++) {
      if (grid[pos[0]][i] === num && pos[1] !== i) {
        return false;
      }
      if (grid[i][pos[1]] === num && pos[0] !== i) {
        return false;
      }
    }

    // check if the number is already in the 3x3 box
    const boxStartRow = pos[0] - (pos[0] % 3);
    const boxStartCol = pos[1] - (pos[1] % 3);

    // iterate through the 3x3 box
    for (let i = boxStartRow; i < boxStartRow + 3; i++) {
      for (let j = boxStartCol; j < boxStartCol + 3; j++) {
        if (grid[i][j] === num && (i !== pos[0] || j !== pos[1])) {
          return false;
        }
      }
    }

    return true; // valid placement
  }

  // function to fill the grid with numbers
  recursiveFillGrid(grid = this.grid) {
    // iterate through the grid to fill the empty cells
    for (let i = 0; i < 9; i++) {
      for (let j = 0; j < 9; j++) {
        if (grid[i][j] === 0) {
          const numbers = Array.from({ length: 9 }, (_, index) => index + 1);
          this.shuffleArray(numbers); // shuffle numbers for new random puzzle
          for (let num of numbers) {
            if (this.isValidPlacement(grid, num, [i, j])) {
              grid[i][j] = num;
              if (this.recursiveFillGrid(grid)) {
                return true;
              }
              grid[i][j] = 0;
            }
          }
          return false; // no valid number for this cell
        }
      }
    }
    return true; // grid is filled
  }

  // wrapper function to generate a complete grid
  generateCompleteGrid() {
    this.initialiseGrid(); // initialise the grid
    if (!this.recursiveFillGrid()) {
      throw new Error("Failed to generate a complete Sudoku grid");
    }
  }

  // function to get the non-empty squares for solution uniqueness check
  getNonEmptySquares() {
    const nonEmptySquares = []; // array to store non-empty squares
    for (let row = 0; row < 9; row++) {
      for (let col = 0; col < 9; col++) {
        if (this.grid[row][col] !== 0) {
          nonEmptySquares.push([row, col]);
        }
      }
    }
    return nonEmptySquares;
  }

  // function to remove numbers from the grid, creating a unique puzzle
  removeNumbersFromGrid() {
    // numbers to remove based on selected difficulty level
    const removalsBasedOnLevel = this.difficultyLevels[this.level];
    let numbersRemoved = 0;

    // remove numbers until the target based on the level is reached
    while (numbersRemoved < removalsBasedOnLevel) {
      // select a random row and column
      let row = Math.floor(Math.random() * 9);
      let col = Math.floor(Math.random() * 9);
      // check if cell is empty
      if (this.grid[row][col] !== 0) {
        // creates a backup of the current grid state before removing a number
        const backupGrid = JSON.parse(JSON.stringify(this.grid));
        this.grid[row][col] = 0; // attempt to remove a number

        // check if current grid config has unique solution
        if (!this.hasUniqueSolution(JSON.parse(JSON.stringify(this.grid)))) {
          this.grid = backupGrid; // restore if uniqueness is compromised
        } else {
          // add to counter, number successfully removed
          numbersRemoved++;
        }
      }
    }
  }

  // function to check if the generated puzzle after removing numbers has unique solution
  hasUniqueSolution(grid) {
    let solutions = 0; //  counter for solutions

    // recursive function to solve the puzzle
    const solve = (grid, currentPos = 0) => {
      // stop if more than one solution is found
      if (solutions > 1) return;
      // base case: if all cells are processed and a solution is found
      if (currentPos === 81) {
        solutions++;
        return;
      }

      // calculate current row and column based on the current position
      const row = Math.floor(currentPos / 9);
      const col = currentPos % 9;

      // move to the next cell if current cell is not empty
      if (grid[row][col] !== 0) {
        solve(grid, currentPos + 1);
        return;
      }
      // try placing numbers 1-9 in the current empty cell
      for (let num = 1; num <= 9 && solutions <= 1; num++) {
        // stop trying numbers if solution is found
        if (this.isValidPlacement(grid, num, [row, col])) {
          grid[row][col] = num; // place num in the grid
          solve(grid, currentPos + 1); // move to solve the next cell
          grid[row][col] = 0; // backtrack by removing the number if it doesn't lead to a solution
        }
      }
    };
    // create a deep copy of the grid to solve without changing original grid
    solve(JSON.parse(JSON.stringify(grid)));
    // true if only one solution is found
    return solutions === 1;
  }

  // print the grid
  printGrid(message = "Grid") {
    console.log(`${message}:`);
    for (let row = 0; row < 9; row++) {
      if (row % 3 === 0 && row !== 0) {
        console.log("- - - - - - - - - - - -");
      }
      let rowString = "";
      for (let col = 0; col < 9; col++) {
        rowString +=
          (col % 3 === 0 && col !== 0 ? " | " : "") + this.grid[row][col] + " ";
      }
      console.log(rowString.trim());
    }
    console.log("\n");
  }

  getGrid() {
    return JSON.parse(JSON.stringify(this.grid));
  }

  // function to shuffle numbers for the number placement
  shuffleArray(array) {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [array[i], array[j]] = [array[j], array[i]];
    }
  }
}

module.exports = SudokuGenerator;

async function main() {
  const sudoku = new SudokuGenerator("Hard"); // adjust to test
  sudoku.generateCompleteGrid();
  sudoku.printGrid("Fully Generated Grid");

  sudoku.removeNumbersFromGrid();
  sudoku.printGrid("Puzzle Grid");

  if (sudoku.hasUniqueSolution(JSON.parse(JSON.stringify(sudoku.grid)))) {
    console.log("A unique solution exists.\n");
  } else {
    console.log("No unique solution exists.");
  }
}

// this ensures that the main function is only called when the file is run directly
if (typeof module !== "undefined" && require.main === module) {
  main();
}
